### PR TITLE
Fix live update consistency in pixel classification

### LINKED ISSUE #2092: Pixel Classification: live update not synchronized among lanes
## Describe the bug
In pixel classification, when working with multiple images, _live update_ used to be synchronized along lanes. Right now the lanes behave weird and independently. One can also arrive at inconsistent states where _live update_ appears to be pressed but one can still go back to feature selection (see screenshot)

## Expected behavior
Pressing _live update_ in __any__ lane should update it globally.

## To Reproduce
Within the Pixel Classification workflow
1. Add multiple images
2. Select some features in the _Feature Selection Applet_
3. In training, do some annotations, click _live update_ mode
4. Switch a lane, see that _live update_ is disabled

## Screenshots
blah

## Desktop (please complete the following information):
 - ilastik version: this problem was introduced in `1.3.3b2`

the cause is most likely somewhere in these two commits:
* 52bd155bc7524d2a7dd3378388db44d896a1f7bd
* f2539169463212a45f59f75bfde0e5d2757b20bf

### PR DESCRIPTION
**Pull request by k-dominik**
Synchronizes the status of live update button and the associated prediction cache across lanes.

Also added a test to define this behavior.

fixes #2092

### CODE REVIEW

File: ilastik/applets/pixelClassification/pixelClassificationGui.py [hunk 1]
```diff
@@ -784,28 +800,53 @@ def hasFeatures(self) -> bool:
...
+        self.labelingDrawerUi.liveUpdateButton.setChecked(enable)
+        self.labelingDrawerUi.liveUpdateButton.blockSignals(False)
+
+        self._update_aux_buttons(enable)
+        # Notify the workflow that some applets may have changed state now.
+        # (For example, the downstream pixel classification applet can
+        #  be used now that there are features selected)
+        self.parentApplet.appletStateUpdateRequested()
+
+    def _update_freeze_status(self, freeze: bool):
```
Comment (emilmelnikov) [inline ilastik/applets/pixelClassification/pixelClassificationGui.py]: Could inline the argument since `freeze == not self.isLiveUpdateEnabled()`.

File: ilastik/applets/pixelClassification/pixelClassificationGui.py [hunk 2]
```diff
@@ -784,28 +800,53 @@ def hasFeatures(self) -> bool:
     def isLiveUpdateEnabled(self):
         return self.labelingDrawerUi.liveUpdateButton.isChecked()
 
-    def setLiveUpdateEnabled(self, checked: bool):
-        if checked:
+    def _update_live_update_button(self, enable: bool):
```
Comment (emilmelnikov) [inline ilastik/applets/pixelClassification/pixelClassificationGui.py]: Could inline argument since `enable == not self.topLevelOperatorView.FreezePredictions.value`.

File: ilastik/applets/pixelClassification/pixelClassificationGui.py [hunk 3]
```diff
@@ -493,7 +491,25 @@ def __init__(self, parentApplet, topLevelOperatorView, labelingDrawerUiPath=None
...
+        # need to initialize the aux buttons once in case _update_live_update_button did not 
+        # result in an update
+        self._update_aux_buttons(not self.topLevelOperatorView.FreezePredictions.value)
+
+        def sync_live_update_with_freeze_value():
+            self._update_live_update_button(not self.topLevelOperatorView.FreezePredictions.value)
+
+        # listen to freezePrediction changes
+        self.topLevelOperatorView.FreezePredictions.notifyDirty(bind(sync_live_update_with_freeze_value))
+        self.__cleanup_fns.append(partial(
```
Comment (emilmelnikov) [inline ilastik/applets/pixelClassification/pixelClassificationGui.py]: `notifyDirty` already returns an unsubscribe callback, so one could use it directly.

File: ilastik/applets/pixelClassification/pixelClassificationGui.py [hunk 4]
```diff
@@ -784,28 +800,53 @@ def hasFeatures(self) -> bool:
...
-    def setLiveUpdateEnabled(self, checked: bool):
-        if checked:
+    def _update_live_update_button(self, enable: bool):
+        button_state = self.isLiveUpdateEnabled()
+        if button_state == enable:
+            return
+
+        if enable:
             if not self.hasFeatures():
                 self.labelingDrawerUi.liveUpdateButton.blockSignals(True)
```
Comment (emilmelnikov) [inline ilastik/applets/pixelClassification/pixelClassificationGui.py]: These 3 lines are redundant since `enable == True` implies `button_state == False` which means an already unchecked button.

File: ilastik/applets/pixelClassification/pixelClassificationGui.py (hunk 2 above)
Comment (emilmelnikov) [reply ilastik/applets/pixelClassification/pixelClassificationGui.py]: Not sure whether we could do any better, but this method's name is awkward...

Comment (k-dominik) [reply ilastik/applets/pixelClassification/pixelClassificationGui.py]: > Not sure whether we could do any better, but this method's name is awkward...

oh wow, i didn't even realize how bad the name was; thx ;)

File: ilastik/applets/pixelClassification/pixelClassificationGui.py (hunk 4 above)
Comment (k-dominik) [reply ilastik/applets/pixelClassification/pixelClassificationGui.py]: good catch!

**Comment by k-dominik** [comment]
> I find it a bit hard to follow the changes, mainly because there are multiple `update_*` functions for FreezePredictions and liveUpdateButton, which should always be in sync (?)
> 
> Maybe we could define a single function that gets called on either FreezePredictions or liveUpdateButton change, and work out all state changes there?

Hm, I think there have to be at least two functions, one that updates the button from the `FreezePrediction` change, and one that updates `FreezePredictions` when the button is toggled. I don't think reworking this into one (as it was before) makes it simpler (didn't work properly before).

**Comment by Tomaz-Vieira** [comment]
Nice that you managed to catch that, even after all the messing around I did with the code =)

So, it seems that the most important difference in this patch is that the applet is watching `FreezePredictions` for dirtyness, and updating accordingly. I naively assumed that there would be only a single applet for all lanes, so it would be ok to initialize it with "live update" disabled. But that meant that when we switched lanes, a new applet would be created, with "live update" disabled, even if it had been enabled in the first lane.

Here's a [patch](https://github.com/ilastik/ilastik/compare/master...Tomaz-Vieira:fix_2096?expand=1) that tries to merge both ideas; it keeps the new code but also watches for `FreezePredictions` dirtyness. Still, if you don't find it to be readable, feel free to just ignore it and carry own with your version.

**Comment by k-dominik** [comment]
Hey @Tomaz-Vieira,

thanks a lot for checking it out! Without a doubt your solution is simpler, I like that :) So I just cherry picked :)
